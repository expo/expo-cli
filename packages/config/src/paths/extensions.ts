import invariant from 'invariant';

export type LanguageOptions = {
  isTS: boolean;
  isModern: boolean;
  isReact: boolean;
};

export function getExtensions(
  platforms: string[],
  extensions: string[],
  workflows: string[]
): string[] {
  // In the past we used spread operators to collect the values so now we enforce type safety on them.
  invariant(Array.isArray(platforms), 'Expected: `platforms: string[]`');
  invariant(Array.isArray(extensions), 'Expected: `extensions: string[]`');
  invariant(Array.isArray(workflows), 'Expected: `workflows: string[]`');

  const fileExtensions = [];
  // support .expo files
  for (const workflow of [...workflows, '']) {
    // Ensure order is correct: [platformA.js, platformB.js, js]
    for (const platform of [...platforms, '']) {
      // Support both TypeScript and JavaScript
      for (const extension of extensions) {
        fileExtensions.push([platform, workflow, extension].filter(Boolean).join('.'));
      }
    }
  }
  return fileExtensions;
}

export function getLanguageExtensionsInOrder({
  isTS,
  isModern,
  isReact,
}: LanguageOptions): string[] {
  // @ts-ignore: filter removes false type
  const addLanguage = (lang: string): string[] => [lang, isReact && `${lang}x`].filter(Boolean);

  // Support JavaScript
  let extensions = addLanguage('js');

  if (isModern) {
    extensions.unshift('mjs');
  }
  if (isTS) {
    extensions = [...addLanguage('ts'), ...extensions];
  }

  return extensions;
}

export function getManagedExtensions(
  platforms: string[],
  languageOptions: LanguageOptions = { isTS: true, isModern: true, isReact: true }
): string[] {
  const fileExtensions = getExtensions(platforms, getLanguageExtensionsInOrder(languageOptions), [
    'expo',
  ]);
  // Always add these last
  _addMiscellaneousExtensions(fileExtensions);
  return fileExtensions;
}

export function getBareExtensions(
  platforms: string[],
  languageOptions: LanguageOptions = { isTS: true, isModern: true, isReact: true }
): string[] {
  const fileExtensions = getExtensions(
    platforms,
    getLanguageExtensionsInOrder(languageOptions),
    []
  );
  // Always add these last
  _addMiscellaneousExtensions(fileExtensions);
  return fileExtensions;
}

function _addMiscellaneousExtensions(fileExtensions: string[]): string[] {
  // Always add these with no platform extension
  // In the future we may want to add platform and workspace extensions to json.
  fileExtensions.push('json');
  fileExtensions.push('wasm');
  return fileExtensions;
}
