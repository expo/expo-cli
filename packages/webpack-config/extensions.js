module.exports = require('./webpack/extensions');
