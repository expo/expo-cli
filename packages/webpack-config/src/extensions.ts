export { default as withUnimodules } from './withUnimodules';
export { default as withWorkbox } from './withWorkbox';
