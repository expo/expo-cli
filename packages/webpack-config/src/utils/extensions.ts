export function getModuleFileExtensionsWithoutDotPrefix(...platforms: string[]): string[] {
  let fileExtensions = [];

  // Support both TypeScript and JavaScript
  for (const extension of ['ts', 'tsx', 'js', 'jsx']) {
    // Ensure order is correct: [platformA.js, platformB.js, js]
    for (const platform of [...platforms, '']) {
      fileExtensions.push([platform, extension].filter(Boolean).join('.'));
    }
  }
  // Always add this last
  fileExtensions.push('json');

  return fileExtensions;
}

export function getModuleFileExtensions(...platforms: string[]): string[] {
  // Webpack requires a `.` before each value
  return getModuleFileExtensionsWithoutDotPrefix(...platforms).map(value => `.${value}`);
}
