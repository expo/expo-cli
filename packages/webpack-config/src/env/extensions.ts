import { getManagedExtensions } from '@expo/config/paths';

/**
 * Get the platform specific platform extensions in the format that Webpack expects (with a dot prefix).
 *
 * @param platforms supported platforms in order of priority. ex: ios, android, web, native, electron, etc...
 * @category env
 */
export function getModuleFileExtensions(...platforms: string[]): string[] {
  // Webpack requires a `.` before each value
  return getManagedExtensions(platforms).map(value => `.${value}`);
}
