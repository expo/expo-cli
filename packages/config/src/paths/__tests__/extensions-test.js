import { getExtensions, getLanguageExtensionsInOrder, getManagedExtensions } from '../extensions';

describe('getExtensions', () => {
  it(`enforces \`string[]\``, async () => {
    expect(() => getExtensions('web')).toThrow(`string[]`);
    expect(() => getExtensions([], 'js')).toThrow(`string[]`);
    expect(() => getExtensions([], [], 'expo')).toThrow(`string[]`);
    expect(() => getExtensions([], null, [])).toThrow(`string[]`);
  });
});

describe('getManagedExtensions', () => {
  it(`creates extensions for web`, async () => {
    expect(getManagedExtensions(['web'])).toMatchSnapshot();
  });

  it(`creates extensions for iOS`, async () => {
    expect(getManagedExtensions(['ios', 'native'])).toMatchSnapshot();
  });
});

// Enforce that all extensions are returned in the correct order, this is very important!
describe('getLanguageExtensionsInOrder', () => {
  // Return only support for the bare minimum
  it(`JS only`, async () => {
    expect(
      getLanguageExtensionsInOrder({ isTS: false, isModern: false, isReact: false })
    ).toStrictEqual(['js']);
  });

  // Modern should come before JS in a basic MJS project
  it(`Modern JS`, async () => {
    expect(
      getLanguageExtensionsInOrder({ isTS: false, isModern: true, isReact: false })
    ).toStrictEqual(['mjs', 'js']);
  });

  // TypeScript should precede JavaScript
  it(`TypeScript`, async () => {
    expect(
      getLanguageExtensionsInOrder({ isTS: true, isModern: false, isReact: false })
    ).toStrictEqual(['ts', 'js']);
  });

  // JS should precede JSX files
  it(`React`, async () => {
    expect(
      getLanguageExtensionsInOrder({ isTS: false, isModern: false, isReact: true })
    ).toStrictEqual(['js', 'jsx']);
  });

  // Modern should come before JS in a basic MJS project
  it(`TypeScript React`, async () => {
    expect(
      getLanguageExtensionsInOrder({ isTS: true, isModern: false, isReact: true })
    ).toStrictEqual(['ts', 'tsx', 'js', 'jsx']);
  });

  // Should include all extensions in the correct order
  it(`Modern TypeScript React`, async () => {
    expect(
      getLanguageExtensionsInOrder({ isTS: true, isModern: true, isReact: true })
    ).toStrictEqual(['ts', 'tsx', 'mjs', 'js', 'jsx']);
  });
});
