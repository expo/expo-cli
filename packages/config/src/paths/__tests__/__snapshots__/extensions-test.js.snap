// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getManagedExtensions creates extensions for iOS 1`] = `
Array [
  "ios.expo.ts",
  "ios.expo.tsx",
  "ios.expo.mjs",
  "ios.expo.js",
  "ios.expo.jsx",
  "native.expo.ts",
  "native.expo.tsx",
  "native.expo.mjs",
  "native.expo.js",
  "native.expo.jsx",
  "expo.ts",
  "expo.tsx",
  "expo.mjs",
  "expo.js",
  "expo.jsx",
  "ios.ts",
  "ios.tsx",
  "ios.mjs",
  "ios.js",
  "ios.jsx",
  "native.ts",
  "native.tsx",
  "native.mjs",
  "native.js",
  "native.jsx",
  "ts",
  "tsx",
  "mjs",
  "js",
  "jsx",
  "json",
  "wasm",
]
`;

exports[`getManagedExtensions creates extensions for web 1`] = `
Array [
  "web.expo.ts",
  "web.expo.tsx",
  "web.expo.mjs",
  "web.expo.js",
  "web.expo.jsx",
  "expo.ts",
  "expo.tsx",
  "expo.mjs",
  "expo.js",
  "expo.jsx",
  "web.ts",
  "web.tsx",
  "web.mjs",
  "web.js",
  "web.jsx",
  "ts",
  "tsx",
  "mjs",
  "js",
  "jsx",
  "json",
  "wasm",
]
`;
